File: backend/app/schemas/checklist.py
"""
Checklist schemas for API request/response validation.
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field


class ChecklistItemBase(BaseModel):
    """Base checklist item schema."""
    item_type: str = Field(..., description="Item type (PPE_HELMET, PPE_MASK, FIRE_EXTINGUISHER)")
    description: str = Field(..., max_length=200, description="Item description")


class ChecklistItemCreate(ChecklistItemBase):
    """Schema for creating a checklist item."""
    pass


class ChecklistItemResponse(ChecklistItemBase):
    """Schema for checklist item response."""
    id: int
    checklist_id: int
    is_checked: bool
    auto_checked: bool
    checked_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class ChecklistItemUpdate(BaseModel):
    """Schema for updating a checklist item."""
    is_checked: bool


class ChecklistBase(BaseModel):
    """Base checklist schema."""
    name: str = Field(..., min_length=1, max_length=100, description="Checklist name")


class ChecklistCreate(ChecklistBase):
    """Schema for creating a checklist."""
    camera_id: int
    items: List[ChecklistItemCreate] = Field(default_factory=list)


class ChecklistResponse(ChecklistBase):
    """Schema for checklist response."""
    id: int
    camera_id: int
    is_active: bool
    items: List[ChecklistItemResponse]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/lib/models/checklist.dart
/// Checklist item model
class ChecklistItem {
  final int id;
  final int checklistId;
  final String itemType;
  final String description;
  final bool isChecked;
  final bool autoChecked;
  final DateTime? checkedAt;
  final DateTime createdAt;

  ChecklistItem({
    required this.id,
    required this.checklistId,
    required this.itemType,
    required this.description,
    required this.isChecked,
    required this.autoChecked,
    this.checkedAt,
    required this.createdAt,
  });

  factory ChecklistItem.fromJson(Map<String, dynamic> json) {
    return ChecklistItem(
      id: json['id'],
      checklistId: json['checklist_id'],
      itemType: json['item_type'],
      description: json['description'],
      isChecked: json['is_checked'],
      autoChecked: json['auto_checked'],
      checkedAt: json['checked_at'] != null
          ? DateTime.parse(json['checked_at'])
          : null,
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  ChecklistItem copyWith({bool? isChecked, bool? autoChecked}) {
    return ChecklistItem(
      id: id,
      checklistId: checklistId,
      itemType: itemType,
      description: description,
      isChecked: isChecked ?? this.isChecked,
      autoChecked: autoChecked ?? this.autoChecked,
      checkedAt: checkedAt,
      createdAt: createdAt,
    );
  }
}

/// Checklist model
class Checklist {
  final int id;
  final int cameraId;
  final String name;
  final bool isActive;
  final List<ChecklistItem> items;
  final DateTime createdAt;
  final DateTime updatedAt;

  Checklist({
    required this.id,
    required this.cameraId,
    required this.name,
    required this.isActive,
    required this.items,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Checklist.fromJson(Map<String, dynamic> json) {
    return Checklist(
      id: json['id'],
      cameraId: json['camera_id'],
      name: json['name'],
      isActive: json['is_active'],
      items: (json['items'] as List)
          .map((item) => ChecklistItem.fromJson(item))
          .toList(),
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  int get checkedCount => items.where((item) => item.isChecked).length;
  int get totalCount => items.length;
  double get progress => totalCount > 0 ? checkedCount / totalCount : 0.0;
}
